src/app/clusters/temperature-control-server/temperature-control-server.cpp

```
@@ -0,0 +1,207 @@
+
+            emberAfStatus = MaxTemperature::Get(endpoint, &maxTemperature);
+            if (emberAfStatus != EMBER_ZCL_STATUS_SUCCESS)
+            {
+                status = app::ToInteractionModelStatus(emberAfStatus);
+                goto exit;
+            }
+
+            if (TemperatureControlHasFeature(endpoint, TemperatureControl::Feature::kTemperatureStep))
+            {
+                int16_t step  = 0;
+                emberAfStatus = Step::Get(endpoint, &step);
+                if (emberAfStatus != EMBER_ZCL_STATUS_SUCCESS)
+                {
+                    status = app::ToInteractionModelStatus(emberAfStatus);
+                    goto exit;
+                }
+
+                if ((targetTemperature.Value() - minTemperature) % step != 0)
+                {
+                    status = Status::ConstraintError;
+                    goto exit;
+                }
+            }
+            else
+            {
+                if (targetTemperature.Value() < minTemperature || targetTemperature.Value() > maxTemperature)
+                {
+                    status = Status::ConstraintError;
+                    goto exit;
+                }
+            }
+            emberAfStatus = TemperatureSetpoint::Set(endpoint, targetTemperature.Value());
+            if (emberAfStatus != EMBER_ZCL_STATUS_SUCCESS)
+            {
+                status = app::ToInteractionModelStatus(emberAfStatus);
+            }
+        }
+        else
+        {
+            status = Status::InvalidCommand;
+        }
+    }
+    if (TemperatureControlHasFeature(endpoint, TemperatureControl::Feature::kTemperatureLevel))
+    {
+        if (targetTemperatureLevel.HasValue())
+        {
+            TemperatureControl::SupportedTemperatureLevelsIterator * instance =
+                TemperatureControl::SupportedTemperatureLevelsIterator::GetInstance();
+            if (instance == nullptr)
+            {
+                status = Status::InvalidCommand;
```

The command is fine.  The problem is with our implementation.  This should return an error code that accurately reflects that.

---

Done